Add the Directus Marketplace

rijkvanzanten opened the pull request:
## Scope

What's changed:

- Added a new Marketplace section in settings that lets you browse and install extensions from our new Extensions Registry
- Added the ability to uninstall extensions from Settings
- Standardized the extensions folder by dropping the deprecated folder-per-type method
- Added a new migrations path setting to move migrations out of the extensions folder

## Potential Risks / Drawbacks

- It's mostly new code, so that's good. That being said..:
- This will allow you to download and run third party code which obviously comes with it's own risks

## Review Notes / Questions

- Please divert your gaze to https://github.com/orgs/directus/projects/26/views/1 for the up to date requirements list to track the status of the want-to-haves.
- Please note any and everything in the thread down below, and I'll put it on the requirements list.
- A lot of extensions currently present in the marketplace are outdated / deprecated. We should make sure we go through and mark the ones we know are good.

Closes #19800, closes #21540

rijkvanzanten committed:
Add support for MIGRATIONS_PATH (#20627)

rijkvanzanten committed:
Merge branch 'main' into next

rijkvanzanten committed:
Merge branch 'main' into next

rijkvanzanten committed:
Merge branch 'main' into next

rijkvanzanten committed:
Deprecate "local" extension types (#20624)

* No longer resolve local non-package extensions

* Remove local extensions from watcher

* Install fs-extra for app dev

* Replace usage of get-extensions

* Rename create functions to match behavior

* Drop extensions prefix from cli create

* Drop name requirement from cli add

* Drop extensions name regex checks

* Don't rely on regex for extension identification

* Add changeset

* Fix build for @directus/extensions

* Don't ensure nested dirs

* Load npm dependency by name prefix

* Throw more helpful error when pkg json is missing

* Fix dependency loading paths

* Only load local extensions if extensions path exists

* Fix merge conflict

* Remove unused fs-extra

* Only sync extensions if location is defined

* `extensions` -> `localExtensions`

* Fix resolving of package extensions

- consistent usage of term "package extensions"
- enhance error messages

* Fix syncing extension from storage

* Revert-revert change from #20627

---------

Co-authored-by: ian <licitdev@gmail.com>
Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

paescuj committed:
Add back `MIGRATIONS_PATH` to `@directus/env` (#21053)

rijkvanzanten committed:
Merge branch 'main' into next

rijkvanzanten committed:
Merge branch 'main' into next

rijkvanzanten committed:
Merge branch 'main' into next

rijkvanzanten committed:
Setup browse endpoint for browsing Directus extensions (#21033)

* Rename tag directus-custom to directus-extension

* Devil's in the details

* [WIP] Start on search lib

* Setup extensions-registry package

* Update pnpm lock

* Move registry search fn to new package

* Squash bugs

* Remove unused deps

* Remove unused dep in `extensions`

* Add tests for validate-limit

* Finish test coverage

* Remove unused constant

* Add back lodash-es 😇

* Install extensions-registry

* Add default keywords

* Expose registry endpoint in rest api

* Add describe module

* Reinstal pnpm

* Drop reliance on author field

The problem is that there's no requirement for it to be a valid(ated) user. Publisher is the only field we can rely on being an existing NPM user.

* Update convertSearchResult test after author change

* Format

* Allow anything after the route

To support packages with scopes

* Add test for describe

* Add more tests

* Throw unprocessablecontenterror instead of error

* Install errors pkg

* Finish test coverage

* Don't require description

* Don't wrap individual keywords

* Run formatter

* Add changeset

* Update packages/extensions-registry/src/modules/search/utils/validate-text.ts

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

* Return sandbox information

* Drop support for directus-custom keyword

* Set type to null for multiple tags

* Restart api for new registry

* Add version utility

* Add out-of-date error

* Add assertion for api versions

* Finish new list method

* Add describe

* Use updated registry functions

* Fix controller usage

* Return output in data flag to be consistent

* Add tests for the describe module

* Add tests for list module

* Finish tests for extensions-registry package

* Remove vscode shenanigans

---------

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

rijkvanzanten committed:
Project Forum: Browse UX (#21276)

* Render marketplace route in-app

* Update api list integration

* Add support for limit/offset in list

* Boilerplate route for drawer

* Render extension readme in drawer

* Render extension detail as separate page

* Setup markup for extension banner

* Render readme with markdown styling

* Boilerplate metadata block

* Use v-list for extension metadata metrics

* Render compatibility metric

* Show downloads and publish date

* Refactor to individual files

* Fix missing import

* Add size metric

* Show author on metadat

* Show verified badge

* Finish metadata

* Run formatter

rijkvanzanten committed:
Style forum detail banner (#21290)

* Style banner contents

* Add background art

* Add banner styles to default dark mode theme

* Undo base style change to avoid conflict

rijkvanzanten committed:
Merge branch 'main' into next

rijkvanzanten committed:
Add beta chip badge to settings nav (#21296)

rijkvanzanten committed:
Add layout styling for detail page (#21297)

* Render metadata as grid if size allows

* Finish layout for detail page

rijkvanzanten committed:
Fix alignment of count

rijkvanzanten committed:
Fix search-input active state

rijkvanzanten committed:
Finalize forum header

rijkvanzanten committed:
Add marketplace banner

rijkvanzanten committed:
Style registry filter

rijkvanzanten committed:
Change v-list-item height to min-height

Allows rows for forum registry to grow

rijkvanzanten committed:
Finalize styling for extension list item

rijkvanzanten committed:
Finalize styling for registry

rijkvanzanten committed:
Don't rely on format util

rijkvanzanten committed:
Forum account detail page (#21299)

* Add support for author describe endpoint

* Boilerplate author page

* Add author endpoint

* Update naming to account

* Render github name on registry list

* Render user's name + avatar on detail

* Load account data on account page

* Restructure module

* Render account banner on account page

* Render account metadata

* Finish account detail page

* Run formatter

rijkvanzanten committed:
Add sidebar info detail sections

mahendraHegde committed:
install extensions from registry (#21070)

* added an endpoint to install extensions from registries

* add changeset

* Revert format/lint scripts

* Update pnpm-lock

* Remove unused dependency

* Add download module to extensions-registry abstraction

* Download from extension registry instead of direct npm

* Throw on error

* Install based on version ID rather than name

* Install based on ID in body

Allows us to extend it in the future if need be

* Download to tmp path instead of extensions

* Save to extensions manager flattened

* Use const for package folder name

* Run more in try/catch to avoid server exit

* Remove changeset

* Undo docs updates

We'll do all docs in a single PR :)

---------

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>
Co-authored-by: Rijk van Zanten <rijkvanzanten@me.com>

rijkvanzanten committed:
Manage extensions by directus_extensions primary key instead of npm package name (#21300)

* Add marketplace migration

* Include bundle info

* Rename name to folder

* Store name as folder

* [WIP] Change name to ID

* Update API to use UUIDs for extensions

* Update app to use extensions id

* Run formatter

* Add changeset

* Update generate-app-entrypoint to use updated format

* Fix graphql usage

* Format

rijkvanzanten committed:
Add forum uninstall endpoint (#21301)

* Bootstrap uninstall

* Finish uninstall endpoint

rijkvanzanten committed:
Use shared extensions store state for managing extensions in settings (#21317)

* Use extensionsstore for extensions management

* Use shared store for extensions state and modification

* Remove unused emit

rijkvanzanten committed:
Merge branch 'main' into next

rijkvanzanten committed:
Adds install/uninstall GUI elements for project forum (#21319)

* Use warning color for incompatible

* Styling tweaks

* Transition optional background color

* Add keys for installing

* Add install button to extension detail

* Finish install button

* Add uninstall option to extensions item

* Empty local cached extensions on start

* Add (hopefully) temporary hack to await reload

* Add install / uninstall buttons

* Update lockfile

* Fix bundle entries not rendering properly

* Clean up spinner styling

* Properly wait for reload to be done

* Run formatter

rijkvanzanten committed:
Fix registry name

rijkvanzanten committed:
Fix local extension loading

rijkvanzanten committed:
Render location/company in byline

rijkvanzanten committed:
Rearchitect metadata to unstyled listitems

rijkvanzanten committed:
Add max-width to page container

rijkvanzanten committed:
Only allow admins to install/uninstall extensions

rijkvanzanten committed:
Remove unused link

rijkvanzanten committed:
Disable eslint for map use

rijkvanzanten committed:
Run formatter

br41nslug reviewed (commented):
The reload page navigation guard and notification seem to be broken and forces you to refresh aggressively on each enable/disable without any way to cancel.

Reloading on `main`:

https://github.com/directus/directus/assets/9389634/94c0ee7c-d2a6-4c78-9119-19b0350b4810


Reloading on `next`:


https://github.com/directus/directus/assets/9389634/420a2771-685d-4b09-83f1-b011aedc253f


Having to do a full refresh then scroll back to the right extension to enable/disable the next one is not a great experience.

Pagination on the Marketplace has some room for improvement:
- navigating between pages is neither registered in the URL or in the browser history (so the browser back/forward dont work as expected)
- pagination does not reset when switching categories, if you go to page 10 on "all" extensions and then open "bundles" it will try to load page 10 which does not exist
- nice-to-have, a pagination button to jump directly to the first or last page as you cannot see all pages


https://github.com/directus/directus/assets/9389634/58195e46-76ce-4423-b9bb-a88cfe1c2cdf

DanielBiegler reviewed (changes requested):
Search field rendering errors + clear-value icon needs some padding

[Screencast from 2024-02-07 14-38-44.webm](https://github.com/directus/directus/assets/14810858/9732fbd0-f9f9-4b16-86c9-ea52f16dae53)

DanielBiegler reviewed (changes requested):
An indicator for empty lists would be nice. This is relevant for both

1. When there are no extensions for that extension type
2. When the user filters search results via the search-input and there are no results

| Currently | How about something similar to our empty v-form but of course with better wording? |
|--------|--------|
| ![image](https://github.com/directus/directus/assets/14810858/ba0edc29-e6b8-4421-b852-46aac4ed157e) | ![image](https://github.com/directus/directus/assets/14810858/f5aed078-f68f-4e09-b34c-3a7af825745b) |

DanielBiegler reviewed (changes requested):
Tiny tweaks:

1. Links inside the extension readme markdown should open in a new tab
2. Make said markdown text selectable. I wanted to look up some terminology used in a readme but was unable to select the text which felt bad from a ux standpoint

DanielBiegler reviewed (commented):
Would be neat if we could jump from an extension to its marketplace entry so we can check out the readme again, for example

Maybe something similar to this:

![image](https://github.com/directus/directus/assets/14810858/c86d43ce-2959-44c9-92d3-16ccf972c570)

azrikahar reviewed (commented):
One tiny potential improvement is when an extension shows "Incompatible with your Project", what if we show the semver string as a tooltip on hover? Just as a way to be a tad more informative.

DanielBiegler reviewed (commented):
Similar to @paescuj suggestion of filtering out empty/default READMEs I'd like to also ignore "banner images" for the tagline because I found a couple that will then show up like so which looks bad:

![image](https://github.com/directus/directus/assets/14810858/111c9cfb-ac35-472b-a60f-935db3b6d5b1)

By "banner image" I mean repos similar to ours which features a banner as the first thing in the README, example:

![image](https://github.com/directus/directus/assets/14810858/a2facfa6-b285-46a3-bf11-2ca881256e46)

DanielBiegler reviewed (commented):
Might be good to only start the installation process after a confirmation modal to avoid accidental misclick-installing extensions

rijkvanzanten committed:
Merge branch 'main' into next

rijkvanzanten committed:
Add no-results indicator (#21389)

* Add no-results indicator

* add error state as well

rijkvanzanten committed:
Standardize trailing slash

rijkvanzanten committed:
Standardize pk check

rijkvanzanten committed:
Resolve unnecessary check

rijkvanzanten commented on app/package.json:
```diff
@@ -140,5 +141,9 @@
 		"vue-tsc": "1.8.27",
 		"vuedraggable": "4.1.0",
 		"wellknown": "0.5.0"
+	},
+	"dependencies": {
+		"@directus/extensions-registry": "workspace:*",
+		"semver": "7.5.4"
```
Good one! I keep forgetting everything is "Dev" for the app

rijkvanzanten committed:
Move all deps to dev deps in app

rijkvanzanten commented on packages/utils/node/list-folders.ts:
```diff
@@ -1,14 +1,26 @@
 	for (const file of files) {
 		const filePath = path.join(fullPath, file);
+
+		if (options?.ignoreHidden && file.startsWith('.')) {
+			continue;
+		}
```
Keeping me sharp, love it

rijkvanzanten committed:
Optimize exec order

rijkvanzanten committed:
Pull data from top level package

rijkvanzanten committed:
Redesign filter bar

rijkvanzanten commented:
![CleanShot 2024-02-09 at 16 41 23@2x](https://github.com/directus/directus/assets/9141017/21f07c27-3aeb-4916-b31d-eb9d439b30c8)

Upgraded sorting / search bar to resolve a bunch of the mentioned points 🙂

rijkvanzanten committed:
Reset page count on filter change

rijkvanzanten committed:
Use default padding on top of page

Feels weird now, but there's a theme-update coming that'll make this make sense :)

rijkvanzanten committed:
Show first/last in pagination

rijkvanzanten committed:
Auto-format package name

rijkvanzanten committed:
Use updated registry api return names

rijkvanzanten committed:
Add sorting by total downloads

rijkvanzanten committed:
Add sparkline for downloads

rijkvanzanten committed:
Add a bit of padding

rijkvanzanten committed:
Finalize sparkline styling

rijkvanzanten committed:
Run formatterg

rijkvanzanten committed:
Merge branch 'main' into next

rijkvanzanten committed:
Add sandbox flag in list call

rijkvanzanten committed:
Add MARKETPLACE_TRUST env var

rijkvanzanten committed:
Set sandbox requirement on download

rijkvanzanten committed:
Render list item badge

rijkvanzanten committed:
Fix padding

rijkvanzanten committed:
Show license in gui

rijkvanzanten committed:
Add support for multiple maintainers

rijkvanzanten committed:
Merge branch 'main' into next

rijkvanzanten committed:
Add activity tracking for install

rijkvanzanten committed:
Make incompatibility notice friendlier

rijkvanzanten committed:
Tweak margin of sparkline

rijkvanzanten committed:
Update default CSP headers to include github content

rijkvanzanten committed:
Make extension readme selectable

rijkvanzanten committed:
Consistent icon, add tooltips

rijkvanzanten committed:
Reduce size of install button

rijkvanzanten committed:
Run formatter

rijkvanzanten commented:
OK all! I think we're ready for another round of reviews and merging of this PR 🙂 

There's a couple points I've jotted down as should/could haves for additional PRs that I don't believe are blocking this one, namely:

- Syncing filters/pagination with URL to maintain navigation[^1]
- Adding fallbacks for media / description / readmes
- Adding a link back to the npm page of a package [^2]
- Opening readme links in a new tab[^3]

[^1]: This is kinda a system wide issue on layouts 👀 
[^2]: Packages aren't always pulled from `npm`
[^3]: This is one of those things that can be surprisingly tricky as it's all rendered markdown rather than components.

paescuj commented:
> 2. Packages aren't always pulled from `npm` [↩](#user-content-fnref-2-a3d2511aeb941f1a71b8922e7e9f6e48)

How about a general link to the package in the original registry, if available (be it npm, gh, ...)? Me and @phazonoverload both think it would be quite valuable to have such a link, and particularly also the original name of the package (since we format it differently), so that the origin of the package is clear.

> 3. This is one of those things that can be surprisingly tricky as it's all rendered markdown rather than components. [↩](#user-content-fnref-3-a3d2511aeb941f1a71b8922e7e9f6e48)

Thankfully, the `md` util got us covered 😃
``` 
v-md="{ value: readme, target: '_blank' }"
```

br41nslug reviewed (changes requested):
Accessibility:
- navigating between pages is neither registered in the URL or in the browser history causing the browser buttons to not work correctly and have quite a frustrating experience while searching/browser extensions.
- What properties does search search through? Does this not include author? I am unable to find any of my extensions by either searching for `brainslug` or `br41nslug` while there are some if you manually go through the pages (i did allow non-sandbox) [In the short term it may be better not to have search]
- Relative URLs in README's are not correctly rendered with the github domain (see the test link https://www.npmjs.com/package/directus-extension-websocket-subscribe)

Design:
- [x] The type dropdown is not sizing properly with its scrollbar (optimally this would just show all items on my desktop where there is plenty of room not to need the scroll bar)
![image](https://github.com/directus/directus/assets/9389634/1d6ac0e7-1244-4a45-9658-0ebbadd87f41)
- [x] The search bar looks a tad funky in dark mode
- [ ] Since it can take some time for the interface to update (which i think will only get slower with more extensions) i think it would be good to have some form of loader state

https://github.com/directus/directus/assets/9389634/6fa4926d-314d-43ac-9848-c6170854b9c9


Errors:
- [x] Not exactly sure where this error is rendered but could we add a more friendly fallback here in case of no readme data found, npm gives us the error "This package does not have a README."
![image](https://github.com/directus/directus/assets/9389634/9b837570-825a-41e9-ba74-0685597ab5bf)


Extensions:
- [ ] Something is not playing well with my pre-existing extensions, not exactly sure what caused this but cleared the db table for further tests and that resolved it
![image](https://github.com/directus/directus/assets/9389634/8301166e-ce82-42b1-80ef-78fcf0f8b351)
- [ ] The extension page should have a badge to distinguish 2 extensions with the same name from different sources (`local` vs `registry`)
![image](https://github.com/directus/directus/assets/9389634/46da5478-c90a-4c13-9458-9a99da2fe8b5)
- [x] App extensions do not seem to load while the extension page says enabled, tested with modules and operations
![image](https://github.com/directus/directus/assets/9389634/549ed65e-7a49-4045-8b70-92a64abde7be)
![image](https://github.com/directus/directus/assets/9389634/3e842f43-f1d9-4085-bcd9-32a567e5515d)
![image](https://github.com/directus/directus/assets/9389634/54e21ccb-32af-4a08-87ba-f0a694cb83ff)

rijkvanzanten committed:
Render readme anchors with target blank

rijkvanzanten committed:
Add min width to v-select components

rijkvanzanten committed:
Fix background on search input

rijkvanzanten committed:
Reduce padding on marketplace pages

rijkvanzanten committed:
Fix showing no results

rijkvanzanten commented on app/src/modules/settings/routes/marketplace/routes/registry/components/inline-filter.vue:
```diff
@@ -0,0 +1,132 @@
+	(val) => {
+		search.value = val;
+	},
+	{ debounce: 300, maxWait: 1000 },
```
I think that might've been the maxWait. Removed that, lets see if that solves it! 🤞🏻

rijkvanzanten committed:
Be consistent

rijkvanzanten commented:
@br41nslug 

> What properties does search search through? Does this not include author? I am unable to find any of my extensions by either searching for brainslug or br41nslug while there are some if you manually go through the pages (i did allow non-sandbox) [In the short term it may be better not to have search]

I think this is one of the (classic) issues of Directus itself: The author is a UUID field on the parent package, and the search just uses `?search=` in the Directus API so doesn't search through author name 🤔

rijkvanzanten committed:
Render type filter menu in full height

rijkvanzanten committed:
Clean up compatibility notice

rijkvanzanten committed:
Hide count result when search count is 0

rijkvanzanten committed:
Use v-show instead

rijkvanzanten committed:
Bump no results notice down a bit

rijkvanzanten committed:
Cleanup size of install buttons

rijkvanzanten committed:
Use buttons for install / author

rijkvanzanten committed:
Fix sparkline rendering

rijkvanzanten commented:
> Ah that makes a lot of sense 🤔 should we perhaps do a more targeted filter including relational fields like author (not sure if there are any others that make sense to search through)

We could theoretically add a `?deep[author][name][_contains]=brainslug` but that would return less instead of more results 🤔

rijkvanzanten committed:
Add better loading state for extension detail page

rijkvanzanten committed:
Add spinners

rijkvanzanten committed:
Render proper error message in case readme is missing

rijkvanzanten committed:
Use heart logo for marketplace

rijkvanzanten committed:
Add image border shadow

rijkvanzanten committed:
Add loading banner on account page

rijkvanzanten committed:
Show reload warning as persistent notification

rijkvanzanten committed:
Only show one notification

rijkvanzanten committed:
Update snapshot

rijkvanzanten committed:
Add padding to verified icon

rijkvanzanten committed:
Base extensions id on package id, show installed badge

rijkvanzanten committed:
Sync pagination / search / filter with URL

Special shoutout to Tim

rijkvanzanten committed:
Merge branch 'main' into next

rijkvanzanten committed:
Next max ext (#21541)

* Add limit exceeded error

* Add limit exceeded translation

* Add extensions_limit env var

* Load extensions limit in app

* Don't allow installing unknown versions

* Don't require installed flag

* Return bundled entries in describe

* Update chip to match style

* Prevent install on limit reached

* Run formatter

rijkvanzanten committed:
Fix bundle loading

rijkvanzanten commented:
OK I'm pens-down on this all! No more improvements on features will be added on this branch. The vast majority of improvement suggestions have been implemented. Lets switch our efforts for one more review to make sure everything works as expected and lets ship this sucker! 🚢

br41nslug reviewed (commented):
I think the "Compatible with your Project" text is a bit misleading or creating false expectations. I've installed a bunch of the available extensions from the marketplace and this label seems counter intuitive:

I've installed "compatible" extensions that absolutely broke all extension loading

![image](https://github.com/directus/directus/assets/9389634/aa370b85-1834-4a6d-b45d-04ca34142740)
![Code_PHxlY6qS0F](https://github.com/directus/directus/assets/9389634/9b824963-4e22-4145-bf12-02859591c28d)

And vice versa installed "not verified to be compatible" extensions made for `9.X` that just work

![image](https://github.com/directus/directus/assets/9389634/680d08c2-b8ad-4966-8f13-ebf25a7a0b2c)

The wording used "Compatible **with your Project**" and "Compatibility **not verified**" implies that we're doing more to actually verify extensions than just the simple string version check that's happening.

rijkvanzanten commented:
@br41nslug Any suggestions? It's the way the extension author communicates whether or not they believe their extension is compatible with the project 🤔 There's no way for us to know if it actually is without trying them all

br41nslug commented:
Changing the label to something like `Built for Directus version XXXX` or something along those lines. Or something that does not imply we're actually checking these extensions like the current wording.

br41nslug reviewed (commented):
An old comment seems to have turned into a bug
> The extension page should have a badge to distinguish 2 extensions with the same name from different sources (local vs registry)

Where it used to render both installed versions (which was confusing but accurate):
Previous screenshot:
![image](https://github.com/directus/directus/assets/9389634/49ef542d-d9df-45ad-9bd0-ae542c37b3c1)

Current state:
![image](https://github.com/directus/directus/assets/9389634/301a72c4-e1dc-4603-bd04-633855f00432)

Since we will have 3 different sources/methods of installing extensions (`local`, `npm` and `marketplace`) being able to spot and fix installations having multiple versions of the same extension installed does not feel like a luxury.

DanielBiegler commented on app/src/modules/settings/routes/marketplace/routes/registry/registry.vue:
```diff
@@ -0,0 +1,197 @@
+
+const perPage = 10;
+
+const search = useRouteQuery('search');
+const page = useRouteQuery('page', 1);
+const type = useRouteQuery('type');
+const sort = useRouteQuery<'popular' | 'recent' | 'downloads'>('sort', 'popular');
```
Having state in the URL is really nice now, but the initial problem of not being able to properly use the browsers back button remains.

1. Go to settings
2. Go to marketplace
3. Press onto some pagination page button, lets say page 3
4. Press the back button
5. You will now go back to the settings instead of going back to the first market place page

This is because we dont push the new URL to the page history

rijkvanzanten replied:
Curse and a blessing? 🤔 It would also be super annoying to have everything in the history state cause then you can never click back to get back to the previous page

br41nslug replied:
I look at it differently 😬 And do consider each page in "pagination" to be its own page (or more broadly each unique filter combination of `type`,`page`,`sort`)

> It would also be super annoying to have everything in the history state cause then you can never click back to get back to the previous page

In its current state it is impossible to get back to the previous page because it does not exist in the history while when everything is on the browser history it is a question of going back enough times to get where you wanted to be (it can take a while if you've done a lot of browsing but at least you can get back if wanted).

DanielBiegler replied:
FWIW: I'm in the same camp as @br41nslug - Personally I'd strongly prefer it being in the history :+1:

rijkvanzanten replied:
Can we push stuff in history in Vue without triggering a navigation event (eg reload)? If that's possible I'm willing to concede.

paescuj replied:
I think a direct [`history.pushState`](https://developer.mozilla.org/en-US/docs/Web/API/History/pushState) call should do it.

rijkvanzanten replied:
Spend about an hour or so trying to sync it into `history.pushState` directly, but am getting some very weird behavior. I'm assuming it's because of vue-router and the browser history getting out of sync while manually adjusting the browser history state. Using `vue-router` `route.push` directly doesn't seem to actually push it into the history state. I'm assuming it's because the vue-router is already on the 'current' route when the change happens. It won't add an entry if it's already the page we're on.

Long story short, I'm treating it as a nice to have but not a blocker for this PR. Lets plan on circling back to this later 👍🏻

paescuj committed:
Resolve npm extensions by package manifest (#21478)

rijkvanzanten committed:
Merge branch 'main' into next

rijkvanzanten commented:
> Changing the label to something like Built for Directus version XXXX or Built for v10.4.3 or something along those lines. Im just suggesting we avoid the use of words like "compatible" or "verified" which people may read literally thinking the extension will actually work or was tested.

The goal for the host version is that extension authors can communicate what versions their extension has been tested on / confirmed working. If that value is incorrect for an individual extension, it should be updated on the extension's repo. On day 1, there will be a lot of extensions where this information is out of date, as we've never surfaced it before. I'm expecting that to smoothen out over time as the community develops more extensions

br41nslug commented:
> The goal for the host version is that extension authors can communicate what versions their extension has been tested on / confirmed working.
> On day 1, there will be a lot of extensions where this information is out of date, as we've never surfaced it before.

I agree, but given that we expect this data to be incorrect lets be very clear here that it was the author of the extension that has "tested on / confirmed working" the extension not us Directus.

paescuj committed:
Merge main into next (#21669)

paescuj committed:
Try to solve conflict in 'next' (#21673)

rijkvanzanten commented:
Continued in https://github.com/directus/directus/pull/21674 due to some git shenanigans
